src/commons/crypto/signers/kmip/internal.rs

```diff
@@ -0,0 +1,776 @@
+            }
+            KeyMaterial::TransparentRSAPrivateKey(priv_key) => {
+                if let Some(public_exponent) = priv_key.public_exponent {
+                    util::rsa_public_key_bytes_from_parts(&priv_key.modulus, &public_exponent)?
+                } else {
+                    return Err(SignerError::KmipError(format!(
+                        "Failed to get key material: missing exponent in transparent RSA private key returned by KMIP Get operation for public key with ID '{}'",
+                        public_key_id)));
+                }
+            }
+            _ => {
+                return Err(SignerError::KmipError(format!(
+                    "Failed to get key material: unsupported key material type {:?} returned by KMIP Get operation for public key with ID '{}'",
+                    key_material, public_key_id)));
+            }
+        };
+
+        Ok(rsa_public_key_bytes)
+    }
+
+    pub fn get_public_key_from_id(&self, public_key_id: &str) -> Result<PublicKey, SignerError> {
+        let rsa_public_key_bytes = self.get_rsa_public_key_bytes(public_key_id)?;
+
+        let subject_public_key = bcder::BitString::new(0, rsa_public_key_bytes);
+
+        let subject_public_key_info =
+            bcder::encode::sequence((PublicKeyFormat::Rsa.encode(), subject_public_key.encode()));
+
+        let mut subject_public_key_info_source: Vec<u8> = Vec::new();
+        subject_public_key_info
+            .write_encoded(bcder::Mode::Der, &mut subject_public_key_info_source)
+            .map_err(|err| {
+                SignerError::KmipError(format!(
+                    "Failed to create DER encoded SubjectPublicKeyInfo from constituent parts: {}",
+                    err
+                ))
+            })?;
+
+        let public_key = PublicKey::decode(subject_public_key_info_source.as_slice()).map_err(|err| {
+            SignerError::KmipError(format!(
+                "Failed to create public key from the DER encoded SubjectPublicKeyInfo: {}",
+                err
+            ))
+        })?;
+
+        Ok(public_key)
+    }
+
+    pub fn sign_with_key(
+        &self,
+        private_key_id: &str,
+        algorithm: SignatureAlgorithm,
+        data: &[u8],
+    ) -> Result<Signature, SignerError> {
+        if algorithm.public_key_format() != PublicKeyFormat::Rsa {
+            return Err(SignerError::KmipError(format!(
+                "Algorithm '{:?}' not supported",
+                algorithm.public_key_format()
+            )));
+        }
+
+        let signed = self
+            .with_conn("sign", |conn| conn.sign(&private_key_id, data))
+            .map_err(|err| SignerError::KmipError(format!("Signing failed: {}", err)))?;
+
+        let sig = Signature::new(SignatureAlgorithm::default(), Bytes::from(signed.signature_data));
+
+        Ok(sig)
+    }
+
+    pub fn destroy_key_pair(&self, kmip_key_pair_ids: &KmipKeyPairIds, mode: KeyStatus) -> Result<bool, SignerError> {
```

Comment:
I wonder if destroy_key should be idempotent. I.e. maybe it should be okay if the key does not exist?

Reply:
I think that removing a key which is already removed is probably not an issue, but.. key cleanups need work. Created an issue for it: https://github.com/NLnetLabs/krill/issues/728